### ISSUE TITLE
Nosetests failing on Windows 

### ISSUE THREAD

**Issue by 7hibault** [nilmtk/nilmtk#361]
Hey,

I've tried to install nilmtk to try it out but I couldn't even get started because the nosetests failed. I have followed the instructions for Windows on the project front page and got the following error after running the nosetests command:
```
Unlink of file 'data/energy.h5' failed. Should I try again? (y/n)
```
When I answer yes, I get that message again and when I answer no, the test fails. Here is the complete output:

```
..........................................Unlink of file 'data/energy.h5' f
ailed. Should I try again? (y/n) y
Unlink of file 'data/energy.h5' failed. Should I try again? (y/n) y
Unlink of file 'data/energy.h5' failed. Should I try again? (y/n) y
Unlink of file 'data/energy.h5' failed. Should I try again? (y/n) n
error: unable to unlink old 'data/energy.h5' (Permission denied)
E
======================================================================
ERROR: test suite for <module 'nilmtk' from 'c:\Users\Tib\Documents\NILM\nil
mtk\nilmtk\__init__.pyc'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\Users\Tib\AppData\Local\Continuum\Anaconda\lib\site-packages\nose
\suite.py", line 228, in run
    self.tearDown()
  File "c:\Users\Tib\AppData\Local\Continuum\Anaconda\lib\site-packages\nose
\suite.py", line 351, in tearDown
    self.teardownContext(ancestor)
  File "c:\Users\Tib\AppData\Local\Continuum\Anaconda\lib\site-packages\nose
\suite.py", line 367, in teardownContext
    try_run(context, names)
  File "c:\Users\Tib\AppData\Local\Continuum\Anaconda\lib\site-packages\nose
\util.py", line 470, in try_run
    return func()
  File "c:\Users\Tib\Documents\NILM\nilmtk\nilmtk\__init__.py", line 29, in
teardown_package
    output = subprocess.check_output(cmd, shell=True, cwd=data_dir())
  File "c:\Users\Tib\AppData\Local\Continuum\Anaconda\lib\subprocess.py", li
ne 573, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
CalledProcessError: Command 'git checkout -- c:\Users\Tib\Documents\NILM\nil
mtk\data' returned non-zero exit status 255

----------------------------------------------------------------------
Ran 42 tests in 25.901s

FAILED (errors=1)
Closing remaining open files:c:\Users\Tib\Documents\NILM\nilmtk\data\energy.
h5...done
```

I have checked on the internet what causes git to display this message. It should appear when another process is using the file. Because git bash was the only opened window, I used ProcessExplorer to find out that three processes were using this file: `git.exe`, `cmd.exe` and `python.exe`. Nothing seems wrong then... Is there anyone with the same issue?

Thanks

**Comment by JackKelly** [comment]
I'm afraid I haven't seen this issue before.  A few quick thoughts:

1. Are you sure the permissions are set correctly?  e.g. see [this blog post](http://fan-zf.blogspot.co.uk/2013/08/error-unable-to-unlink-old-permission.html) (that post is for Linux but may still give some clues).
2. Can you try manually deleting the files in `nilmtk/data` and doing a `git pull` again.  Does this produce any errors?
3. [This post](http://stackoverflow.com/a/7199888/732596) says "it can also be related with the way Git has been installed (on Windows, UAC can generate problem if msysgit is installed in C:\Program or C:\Program Files".  See the post for links on how to deal with this.
4. Have you tried everything suggested [on this stack overflow](http://stackoverflow.com/questions/7199496/git-checkout-warning-unable-to-unlink-files-permission-denied) post?

To give a bit of background... NILMTK's tests modify the test data files in `nilmtk/data` so, at the end of the tests, the tests use `git pull` to get the original files back.  This is a bit of an ugly hack to be honest but it usually works.  You could fairly safely modify the tests so they don't run `git pull` (comment out the `def teardown_package() function` in `nilmtk/__init__.py`)

**Comment by oliparson** [comment]
Hmmmm this is very strange. I've done a lot of NILMTK dev on Windows and not seen this problem before.

**Comment by 7hibault** [comment]
@JackKelly 
Thank you for your feedback, here is what I've tried
<ol>
<li>h5 files are -rw-r--r-- ; this might be the issue but I can't get those permissions to change (there is no ch{mod,own,grp} in windows). I tried to modify their rights with icacls but it didn't solve the issue</li>
<li>I've tried doing that already: it gives no error (git pull doesn't download the files again). Then, running nosetests generates an error + the files and then fails in the same way if ran again</li>
<li>Git was installed in Program Files, I've reinstalled it in C:\ but still had the same issue</li>
<li> Pretty much: ownership/rights ; deleting files ; git location ; file lock. The rights modification seems to be the one thing I haven't really managed to try properly.</li>
</ol>
Also, commenting the `teardown_package` function did allow the tests to run OK! It really narrows down the problem to a git thing. Indeed it's a bit of a tricky hack but I still don't see why it wouldn't work. Copying the original files and restoring them might be a safer alternative though.

@oliparson 
We probably have a different setup? Or maybe I did something wrong in the installation process but I can't see what could've caused that error.

**Comment by JackKelly** [comment]
> Copying the original files and restoring them might be a safer alternative though.

Absolutely.  I think that's what we did originally but we ran into an issue which forced us to use the nasty git hack we currently use (the hack is mine so I can be rude about it!).  @nipunreddevil do you remember why we switched away from copying the .h5 files?  I have a faint memory that we had issues on Windows... not sure though...

**Comment by nipunreddevil** [comment]
Even I don't seem to remember! I think other package like sklearn also test on datasets. However, I think they use standard data sets which do not need modification. I am wondering if we need to even generate newer H5 files?

**Comment by JackKelly** [comment]
OK, well, let's leave this issue open for now and at some time we'll write a simple patch to copy files from a backup rather than use git.  @7hibault is this bug preventing you from getting work done or does the work around work for you for the time being?  Also, would you be at all interested in submitting a pull request to implement an alternative mechanism whereby the data files are copied from a backup rather than using git?

**Comment by 7hibault** [comment]
@JackKelly The work around is fine for the time being, if anyone else faces the issue, I'll think of implementing an alternative.

**Comment by bHuquet** [comment]
Just to let you know: I faced the same problem "ERROR: test suite for <module 'nilmtk' from 'c:\Users\Tib\Documents\NILM\nil
mtk\nilmtk\__init__.pyc'>".
In my case, commenting the  teardown_package  function did allow the tests to run OK as for 7hibault.

**Comment by lionelBytes** [comment]
Hey all, wanted to let you know that I exactly the same problem and until I found this thread I was rather confused! After following the installation instructions on Windows and running nosetests I got:

`C:\Users\l\GreenRunning\nilmtk>nosetests
...........................................Unlink of file 'data/energy.h5' failed. Should I try again? (y/n)`

As above I've commented out the <teardown_package():> and the nosetests now run sucessfully
